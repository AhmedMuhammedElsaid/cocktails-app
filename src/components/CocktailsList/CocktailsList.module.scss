@import "../../scss//variables.scss";
.Container {
  display: grid;
  grid-template-columns: repeat(4, 1fr);
  @media (max-width: 1200px) {
    grid-template-columns: repeat(2, 1fr);
  }
  @media (max-width: 600px) {
    grid-template-columns: repeat(1, 1fr);
  }
  grid-gap: 20px;
  background-color: $mainGrey;
}
