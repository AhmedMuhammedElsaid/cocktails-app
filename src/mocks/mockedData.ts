export const cocktailDetailsMockedData = {
  idDrink: "14608",
  strDrink: "Cream Soda",
  strCategory: "Cocktail",
  strAlcoholic: "Alcoholic",
  strGlass: "Highball glass",
  strInstructions:
    "Pour 1oz of Spiced Rum into a highball glass with ice. Fill with Ginger Ale.",
  strInstructionsDE:
    "Geben Sie 3 cl. Spiced Rum in ein Highball-Glas mit Eis. Mit Ginger Ale .",
  strInstructionsIT: `Versare 30gr di rum speziato in un bicchiere highball con ghiaccio.Riempi con Ginger Ale.`,
  strDrinkThumb: `https://www.thecocktaildb.com/images/media/drink/yqstxr1479209367.jpg`,
  strIngredient1: `Spiced rum`,
  strIngredient2: `Ginger ale`,
  strCreativeCommonsConfirmed: `No`,
  dateModified: `2016-11-15 11:29:27`,
};
export const cocktailListMockedData = [
  cocktailDetailsMockedData,
  { ...cocktailDetailsMockedData, idDrink: "14709" },
];
export const cocktailCardMockedData = {
  strDrink: "Cream Soda",
  strCategory: "Cocktail",
  strDrinkThumb: `https://www.thecocktaildb.com/images/media/drink/yqstxr1479209367.jpg`,
  strInstructions:
    "Geben Sie 3 cl. Spiced Rum in ein Highball-Glas mit Eis. Mit Ginger Ale .",
};
