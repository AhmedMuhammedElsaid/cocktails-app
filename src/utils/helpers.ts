/*
 @params object
 return object with no null ,undefined , empty string "cleaned object"
*/
export function cleanObject(object: { [x: string]: any }) {
  if (!object) return {};
  if (typeof object === "object") {
    for (let propName in object) {
      if (
        object[propName] === null ||
        object[propName] === undefined ||
        object[propName]?.length === 0
      ) {
        delete object[propName];
      }
    }
  }
  return object;
}
