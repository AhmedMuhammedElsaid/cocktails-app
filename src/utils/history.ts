import { createBrowserHistory } from "history";

const history = createBrowserHistory();
// method to redirect the user to error page in case any error happens while fetching the data from the server
export const redirectToErrorPage = () => {
  history.push(`/pagenotfound`);
  history.go(0);
};
export default history;
