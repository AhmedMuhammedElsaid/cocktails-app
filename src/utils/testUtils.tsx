import { render, RenderResult } from "@testing-library/react";
import { BrowserRouter as Router } from "react-router-dom";

// this is a handy function that I would utilize for any component
// that relies on the router being in context
export const customRender = (component: JSX.Element): RenderResult => {
  return {
    ...render(<Router>{component}</Router>),
  };
};
