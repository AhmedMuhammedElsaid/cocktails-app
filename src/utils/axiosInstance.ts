import axios from "axios";
import { redirectToErrorPage } from './history';
import { toast } from "react-toastify";

const axiosInstance = axios.create({
  // setting the default uri for this instance
  baseURL: process.env.REACT_APP_BASE_URI,
});

// setting language  configuration on every single request we send to the BE
axiosInstance.interceptors.request.use(
  (request: any) => {
    return {
      ...request,
      headers: {
        ...request.headers,
        // setting the header lang german "de"
        "Accept-Language": "de",
      },
    };
  },
  (error) => {
    return Promise.reject(error);
  }
);

axiosInstance.interceptors.response.use(
  (response) => {
    return response.data;
  },
  (error) => {
    switch (error?.response?.status) {
      case 400:
        return {
          data: null,
          hasError: true,
          error: [error.response.data],
        };
      case 401:
        toast.error("Unauthorized Call Access");
        break;
      case 404:
        toast.error("404 Not Found");
        break;
      case 500:
        toast.error("Server Is Down");
        break;
    }
    redirectToErrorPage()
    return Promise.reject(error);
  }
);
export default axiosInstance;
