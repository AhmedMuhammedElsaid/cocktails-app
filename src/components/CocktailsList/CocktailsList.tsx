import React, { FunctionComponent } from "react";
import CocktailCard from "../CocktailCard/CocktailCard";
import { Grid } from "@material-ui/core";
import { cocktailSchema } from "utils/types";
import styles from "./CocktailsList.module.scss";

interface Props {
  cocktails: cocktailSchema[] | null;
}
const CocktailsList: FunctionComponent<Props> = ({ cocktails }) => {
  return (
    <Grid className={styles.Container} data-testid="card-holder-testid">
      {cocktails?.map((cocktail) => (
        <CocktailCard cocktail={cocktail} key={cocktail.idDrink} />
      ))}
    </Grid>
  );
};
export default CocktailsList;
