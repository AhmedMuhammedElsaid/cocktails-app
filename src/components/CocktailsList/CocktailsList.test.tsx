import { cleanup } from "@testing-library/react";
import "@testing-library/jest-dom";
import CocktailsList from "./CocktailsList";
import { customRender } from "utils/testUtils";
import { cocktailListMockedData } from "mocks/mockedData";

const cocktailListHolder = "cocktail-card-testid";
describe("Test Cocktail List Component", () => {
  afterEach(cleanup);
  it("should render all cocktails data passed from outside", () => {
    const { asFragment, getAllByTestId } = customRender(
      <CocktailsList cocktails={cocktailListMockedData} />
    );

    expect(getAllByTestId(cocktailListHolder)).toHaveLength(2);
    expect(asFragment()).toMatchSnapshot();
  });
  it(`Show no-data component in case didn't pass cocktails data`, () => {
    const { asFragment, queryByTestId } = customRender(<CocktailsList />);
    expect(queryByTestId(cocktailListHolder)).not.toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });
});
