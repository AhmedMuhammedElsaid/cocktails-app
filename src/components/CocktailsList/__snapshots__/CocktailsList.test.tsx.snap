// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Cocktail List Component Show no-data component in case didn't pass cocktails data 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root Container"
    data-testid="card-holder-testid"
  />
</DocumentFragment>
`;

exports[`Test Cocktail List Component should render all cocktails data passed from outside 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root Container"
    data-testid="card-holder-testid"
  >
    <a
      class="MuiPaper-root MuiCard-root Card MuiPaper-elevation1 MuiPaper-rounded"
      data-testid="cocktail-card-testid"
      href="/cocktail-details/14608"
    >
      <button
        class="MuiButtonBase-root MuiCardActionArea-root"
        tabindex="0"
        type="button"
      >
        <img
          alt="Cream Soda"
          class="MuiCardMedia-root Card__Img MuiCardMedia-media MuiCardMedia-img"
          data-testid="cocktail-img-testid"
          height="250"
          src="https://www.thecocktaildb.com/images/media/drink/yqstxr1479209367.jpg"
        />
        <div
          class="MuiCardContent-root"
        >
          <div
            class="MuiTypography-root Cocktail__Name MuiTypography-h6 MuiTypography-gutterBottom"
            data-testid="cocktail-name-testid"
          >
            Cream Soda
          </div>
           
          <div
            class="MuiTypography-root Category MuiTypography-h6 MuiTypography-gutterBottom"
            data-testid="cocktail-category-testid"
          >
            Cocktail
          </div>
          <p
            class="MuiTypography-root Description MuiTypography-body2"
            data-testid="cocktail-description-testid"
          >
            Pour 1oz of Spiced Rum into a highball glass with ice. Fill with Ginger Ale....
          </p>
        </div>
        <span
          class="MuiCardActionArea-focusHighlight"
        />
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </a>
    <a
      class="MuiPaper-root MuiCard-root Card MuiPaper-elevation1 MuiPaper-rounded"
      data-testid="cocktail-card-testid"
      href="/cocktail-details/14709"
    >
      <button
        class="MuiButtonBase-root MuiCardActionArea-root"
        tabindex="0"
        type="button"
      >
        <img
          alt="Cream Soda"
          class="MuiCardMedia-root Card__Img MuiCardMedia-media MuiCardMedia-img"
          data-testid="cocktail-img-testid"
          height="250"
          src="https://www.thecocktaildb.com/images/media/drink/yqstxr1479209367.jpg"
        />
        <div
          class="MuiCardContent-root"
        >
          <div
            class="MuiTypography-root Cocktail__Name MuiTypography-h6 MuiTypography-gutterBottom"
            data-testid="cocktail-name-testid"
          >
            Cream Soda
          </div>
           
          <div
            class="MuiTypography-root Category MuiTypography-h6 MuiTypography-gutterBottom"
            data-testid="cocktail-category-testid"
          >
            Cocktail
          </div>
          <p
            class="MuiTypography-root Description MuiTypography-body2"
            data-testid="cocktail-description-testid"
          >
            Pour 1oz of Spiced Rum into a highball glass with ice. Fill with Ginger Ale....
          </p>
        </div>
        <span
          class="MuiCardActionArea-focusHighlight"
        />
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </a>
  </div>
</DocumentFragment>
`;
